### PR TITLE
Add v4Package Schema

### PR DESCRIPTION
**Pull request by larioj**
We are adding updates to packages. As part of this effort we are adding a new package specification which encodes upgrade information. Specifically we are adding a field that shows for a given package `A`, what other packages `upgrades=[B, C, D]` are able to upgrade to `A`.

### CODE REVIEW

File: repo/meta/schema/package-schema.json [hunk 1]
```diff
@@ -192,14 +192,112 @@
...
+          "type": "boolean",
+          "default": false,
+          "description": "True if this package installs a new Mesos framework."
+        },
+        "upgrades": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "List of versions from which this package can upgrade from. If the property is missing or null it means that it can upgrade from any package. If the property is the empty list then it means that it can’t upgrade from any package."
```
Comment (cruhland) [inline repo/meta/schema/package-schema.json]: Some suggestions:
1. "List of versions that can upgrade to this package."
1. "If the property is missing or null, any version can upgrade to this package."
1. "If the property is the empty array, no version can upgrade to this package."

**Comment by cruhland** [comment]
Also, I assume we're not creating a `version-4.x` branch of the repo this time, since we now have the schema versions?

File: repo/meta/schema/package-schema.json [hunk 2]
```diff
@@ -114,7 +114,8 @@
           "type": "string"
         },
         "version": {
-          "type": "string"
+          "type": "string",
+          "pattern": "^[-a-zA-Z0-9.]+$"
```
Comment (cruhland) [inline repo/meta/schema/package-schema.json]: Did you check that this is compatible with all versions currently in Universe? Or is this just the characters allowed by semver?

File: repo/meta/schema/package-schema.json [hunk 3]
```diff
@@ -192,14 +193,123 @@
...
+          "type": "boolean",
+          "default": false,
+          "description": "True if this package installs a new Mesos framework."
+        },
+        "upgradesFrom": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "List of versions that can upgrade to this package. If the property is a list containing '*', any version can upgrade to this package. If the property is the empty list, no version can upgrade to this package."
```
Comment (cruhland) [inline repo/meta/schema/package-schema.json]: What happens if the list has both version numbers and `*`?

File: scripts/validate-packages.py [hunk 4]
```diff
@@ -57,6 +57,15 @@ def _validate_revision(given_package, revision, path):
...
     packaging_version = package_json.get("packagingVersion", "2.0")
 
+    # validate upgrades version
+    if packaging_version == "4.0":
+        min_dcos_version = package_json.get("minDcosReleaseVersion").split('.')
+        major = int(min_dcos_version[0])
+        minor = int(min_dcos_version[1])
+
+        if major < 1 or (major == 1 and minor < 10):
+            sys.exit("\tERROR\n\nminDcosReleaseVersion must be greater than or equal to 1.10")
```
Comment (cruhland) [inline scripts/validate-packages.py]: Looks like the other checks in this file use `eprint()`?

File: scripts/validate-packages.py [hunk 5]
```diff
@@ -57,6 +57,15 @@ def _validate_revision(given_package, revision, path):
 
     packaging_version = package_json.get("packagingVersion", "2.0")
 
+    # validate upgrades version
+    if packaging_version == "4.0":
+        min_dcos_version = package_json.get("minDcosReleaseVersion").split('.')
+        major = int(min_dcos_version[0])
+        minor = int(min_dcos_version[1])
+
+        if major < 1 or (major == 1 and minor < 10):
```
Comment (cruhland) [inline scripts/validate-packages.py]: Let's follow the existing conventions for comparing release versions in this code: https://github.com/larioj/universe/blob/7465bacd4c151afde330bbfffc70832dd7854f46/scripts/validate-packages.py#L109

**Comment by cruhland** [comment]
@larioj overall looks good, had a few comments.

File: repo/meta/schema/package-schema.json (hunk 2 above)
Comment (larioj) [inline repo/meta/schema/package-schema.json]: I checked that this is compatible with all current v3Package versions. Also the fact that the CI passes proves this.

File: repo/meta/schema/package-schema.json (hunk 3 above)
Comment (larioj) [inline repo/meta/schema/package-schema.json]: ` If the property is a list containing '*', this package can downgrade to any version.` Any list containing star is equivalent to ['*'].

File: scripts/validate-packages.py (hunk 4 above)
Comment (larioj) [inline scripts/validate-packages.py]: The checks that say OK, or some Info use eprint(). The checks that fail the verification use sys.exit()

File: scripts/validate-packages.py (hunk 5 above)
Comment (larioj) [inline scripts/validate-packages.py]: Thanks. That's a lot easier.

File: repo/meta/schema/package-schema.json (hunk 3 above)
Comment (jsancio) [inline repo/meta/schema/package-schema.json]: How about `... If the property is a list containing the string "*", any...`?

File: repo/meta/schema/package-schema.json [hunk 6]
```diff
@@ -192,14 +193,123 @@
...
+      },
+      "required": [
+        "packagingVersion",
+        "name",
+        "version",
+        "maintainer",
+        "description",
+        "tags",
+        "upgradesFrom",
+        "downgradesTo",
```
Comment (jsancio) [inline repo/meta/schema/package-schema.json]: Okay, are we going to convert these packages to an object/package without these two properties (upgradesFrom, downgradTo) if the values are the empty array?

File: scripts/validate-packages.py [hunk 7]
```diff
@@ -57,6 +57,11 @@ def _validate_revision(given_package, revision, path):
 
     packaging_version = package_json.get("packagingVersion", "2.0")
 
+    # validate upgrades version
+    min_dcos_release_version = package_json.get("minDcosReleaseVersion")
+    if packaging_version == "4.0" and LooseVersion(min_dcos_release_version) < LooseVersion("1.10"):
+        sys.exit("\tERROR\n\nminDcosReleaseVersion must be greater than or equal to 1.10")
```
Comment (jsancio) [inline scripts/validate-packages.py]: Can we do better? Can we convert these packages to version 3.0 if both upgradeFrom and downgradeTo are the empty list? What do you think?

Comment (larioj) [inline scripts/validate-packages.py]: Absolutely, I made them optional. If upgradesFrom, and downgradesTo are null, then they are equivalent to the empty list.

Comment (larioj) [inline scripts/validate-packages.py]: Converting V4 to V3 makes sense under these semantics.

File: repo/meta/schema/package-schema.json (hunk 6 above)
Comment (larioj) [inline repo/meta/schema/package-schema.json]: Yes

File: scripts/validate-packages.py (hunk 7 above)
Comment (jsancio) [inline scripts/validate-packages.py]: @larioj Don't we need to update this check after your change?

**Pull request merged**